ekncontent/ekncontent/eknc-query-object.c

```diff
@@ -0,0 +1,829 @@
+   */
+  eknc_query_object_props[PROP_TAGS_MATCH_ALL] =
+    g_param_spec_variant ("tags-match-all", "Tags match all",
+      "A list of tags to restrict the search to",
+      G_VARIANT_TYPE ("as"), NULL,
+      G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  /**
+   * EkncQueryObject:tags-match-any:
+   *
+   * A list of tags to restrict the search to.
+   */
+  eknc_query_object_props[PROP_TAGS_MATCH_ANY] =
+    g_param_spec_variant ("tags-match-any", "Tags match any",
+      "A list of tags to restrict the search to",
+      G_VARIANT_TYPE ("as"), NULL,
+      G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  /**
+   * EkncQueryObject:ids:
+   *
+   * A list of specific ekn ids to limit the search to. Can be used with an
+   * empty query to retrieve the given set of ids.
+   */
+  eknc_query_object_props[PROP_IDS] =
+    g_param_spec_variant ("ids", "Ids",
+      "A list of model ids",
+      G_VARIANT_TYPE ("as"), NULL,
+      G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  /**
+   * EkncQueryObject:excluded-ids:
+   *
+   * A list of specific ekn ids to exclude from the search.
+   */
+  eknc_query_object_props[PROP_EXCLUDED_IDS] =
+    g_param_spec_variant ("excluded-ids", "Excluded ids",
+      "A list of specific ekn ids to exclude from the search",
+      G_VARIANT_TYPE ("as"), NULL,
+      G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  /**
+   * EkncQueryObject:excluded-tags:
+   *
+   * A list of specific ekn tags to exclude from the search.
+   */
+  eknc_query_object_props[PROP_EXCLUDED_TAGS] =
+    g_param_spec_variant ("excluded-tags", "Excluded tags",
+      "A list of specific ekn tags to exclude from the search",
+      G_VARIANT_TYPE ("as"), NULL,
+      G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class,
+                                     NPROPS,
+                                     eknc_query_object_props);
+}
+
+static void
+eknc_query_object_init (EkncQueryObject *self)
+{
+}
+
+// Returns a newly allocated string with the term parenthesized
+static gchar *
+parenthesize (const gchar *term)
+{
+  return g_strdup_printf ("(%s)", term);
+}
+
+// Unicode friendly capitalization function
+static gchar *
+capitalize (const gchar *input)
+{
+  if (input == NULL)
+    return NULL;
+  GString *string = g_string_new (NULL);
+  if (*input)
+    {
+      gunichar firstchar = g_utf8_get_char (input);
+      gunichar titlechar = g_unichar_totitle (firstchar);
+      g_string_append_unichar (string, titlechar);
+      if (g_utf8_next_char (input))
+        g_string_append (string, g_utf8_next_char (input));
+    }
+  return g_string_free (string, FALSE);
+}
+
+// Limit term length we sent to xapian
+static gchar *
+chomp_term (const gchar *term)
+{
+  // Just dupe if the term is short
+  if (strlen (term) <= MAX_TERM_LENGTH)
+    return g_strdup (term);
```

Comment:
Might be a fairly low-cost optimization to have this function operate on the string in-place, so you don't have to allocate a whole new array for the chomped terms. You can just overwrite the first byte of the chomped part with `'\0'` to terminate the string.